src/main/scala/com/gu/newsletterlistcleanse/UpdateBrazeUsersLambda.scala

```diff
@@ -7,73 +7,99 @@ import com.amazonaws.auth.AWSCredentialsProvider
-import com.gu.newsletterlistcleanse.braze.{BrazeClient, BrazeError, BrazeNewsletterSubscriptionsUpdate, BrazeResponse, UserTrackRequest}
+import com.gu.newsletterlistcleanse.braze.{BrazeClient, BrazeError, BrazeNewsletterSubscriptionsUpdate, SimpleBrazeResponse, UserExportRequest, UserTrackRequest}
 import com.gu.newsletterlistcleanse.models.CleanseList
-import io.circe
-import io.circe.parser.decode
+import com.gu.identity.model.EmailNewsletter
+import com.gu.newsletterlistcleanse.sqs.ParseSqsMessage
 import org.slf4j.{Logger, LoggerFactory}
 
-import scala.collection.JavaConverters._
 import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, Future}
+import scala.concurrent.ExecutionContext.Implicits.global
 
 
 class UpdateBrazeUsersLambda {
-
   val logger: Logger = LoggerFactory.getLogger(this.getClass)
 
   val credentialProvider: AWSCredentialsProvider = new NewsletterSQSAWSCredentialProvider()
   val config: NewsletterConfig = NewsletterConfig.load(credentialProvider)
+  val apiKey: String = config.brazeApiToken
 
   val timeout: Duration = Duration(15, TimeUnit.MINUTES)
 
   def handler(sqsEvent: SQSEvent): Unit = {
     val env = Env()
     logger.info(s"Starting $env")
-    parseCleanseListSqsMessage(sqsEvent) match {
+    ParseSqsMessage[CleanseList](sqsEvent) match {
       case Right(cleanseLists) =>
-        process(cleanseLists)
+        resolveBrazeUpdates(cleanseLists)
       case Left(parseErrors) =>
         parseErrors.foreach(e => logger.error(e.getMessage))
     }
-
   }
 
-  def parseCleanseListSqsMessage(sqsEvent: SQSEvent): Either[List[circe.Error], List[CleanseList]] = {
-    (for {
-      message <- sqsEvent.getRecords.asScala.toList
-    } yield {
-      decode[CleanseList](message.getBody)
-    }).toEitherList
+  private def resolveBrazeUpdates(cleanseLists: List[CleanseList]) = {
+    val brazeResults = Await.result(getBrazeResults(cleanseLists), timeout)
+    brazeResults match {
+      case Left(brazeErrors) =>
+        brazeErrors.foreach(e => logger.error(s"Error updating Braze with Code ${e.code}: ${e.body}"))
+      case Right(_) => logger.info("Successfully updated Braze")
+    }
   }
 
-  def process(cleanseLists: List[CleanseList]): Either[List[BrazeError], List[BrazeResponse]] = {
+  private def getInvalidUsers(userIds: List[String]): Future[Either[BrazeError, List[String]]] =
+    BrazeClient.getInvalidUsers(apiKey, UserExportRequest(userIds))
 
-    val brazeResponses = for {
+  private def getAllInvalidUsers(cleanseLists: List[CleanseList]): Future[Either[List[BrazeError], List[String]]] = {
+    val allInvalidUserTasks: List[Future[Either[BrazeError, List[String]]]] = for {
       cleanseList <- cleanseLists
-      newsletterName = cleanseList.newsletterName
-      userIds = cleanseList.userIdList
+      batchedUserIds = cleanseList.userIdList.grouped(50)
+      batch <- batchedUserIds
+    } yield getInvalidUsers(batch)
```

Comment:
Here there might or might not be an issue as the loop will create as many futures as quickly as possible. So depending on the HTTP client, we may hammer the Braze service, or we may not. It might be worth looking into how sttp handles concurrency.

Reply:
So we're using this backend: https://sttp.softwaremill.com/en/v2.0.0-rc6/backends/asynchttpclient.html
What would be a general solution here? Limiting the pool size to make sure that we limit how many can be sent at once?